openedx_events/tooling.py

```diff
@@ -0,0 +1,177 @@
+"""
+Tooling necessary to use Open edX events.
+"""
+import warnings
+from datetime import datetime
+
+import crum
+from django.conf import settings
+from django.dispatch import Signal
+
+import openedx_events
+from openedx_events.exceptions import InstantiationError, SenderValidationError
+
+
+class OpenEdxPublicSignal(Signal):
+    """
+    Custom class used to create Open edX events.
+    """
+
+    def __init__(self, event_type="", data=None, minor_version="0.0"):
+        """
+        Init method for OpenEdxPublicSignal definition class.
+
+        Arguments:
+            event_type (str): name of the event.
+            data (dict): attributes passed to the event.
+            minor_version (str): version of the event type.
+        """
+        if not event_type:
+            raise InstantiationError(
+                message="Missing required argument 'event_type'"
+            )
+        if not data:
+            raise InstantiationError(
+                event_type=event_type, message="Missing required argument 'data'"
+            )
+        self.init_data = data
+        self.event_type = event_type
+        self.minor_version = minor_version
+        super().__init__()
+
+    def __repr__(self):
+        """
+        Represent OpenEdxPublicSignal as a string.
+        """
+        return "<OpenEdxPublicSignal: {event_type}>".format(event_type=self.event_type)
+
+    def get_signal_metadata(self):
+        """
+        Get signal extra metadata when an event is sent.
+
+        These fields are generated on the fly and are a subset of the Event
+        Message defined in the OEP-41.
+
+        Example usage:
+            >>> STUDENT_REGISTRATION_COMPLETED.get_signal_metadata()
+            {
+                'event_type': '...learning.student.registration.completed.v1',
+                'minorversion': '0.0',
+                'time': '2021-06-09T14:12:45.320819Z',
+                'source': 'openedx/lms/web',
+                'sourcehost': None,
+                'specversion': '0.1.0'
+            }
+        """
+
+        def get_current_time():
+            """
+            Getter function used to get timestamp when the event ocurred.
+            """
+            return str(datetime.utcnow().isoformat()) + "Z"
+
+        def get_source():
+            """
+            Getter function used to get logical source of an event.
+            """
+            return "openedx/{service}/web".format(
+                service=getattr(settings, "SERVICE_VARIANT", "")
+            )
+
+        def get_source_host():
+            """
+            Getter function used to get physical source of the event.
+            """
+            current_request = crum.get_current_request()
+            return current_request.get_host() if current_request else None
+
+        def get_spec_version():
+            """
+            Getter function used to get the Open edX Events version.
+            """
+            return openedx_events.__version__
```

Comment:
I like the idea of sending the version of the openedx_events library, but I would not like to mix it with the specversion of CloudEvents.
I'd rather leave `specversion` alone as "1.0" to not mess around and send another field with the openedx events version.

For the new field I'm thinking `events_version`, but it could be other key

Reply:
You could also play off the `source*` convention and call it `sourcelib` or something, but I'm totally fine with `events_version` as well.